### PR TITLE
[15.0][IMP] mrp_multi_level: safety stock

### PR DESCRIPTION
**Pull request by gurneyalex**
When a product has a safety stock in an area, and the current stock is below safety, but there are moves in the future, mrp_multi_level does not compute an immediate action to get back to the safety stock.

This PR changes this behavior: when the mrp_moves are considered, before processing the first move in the future, we insert a resupply action to rebuild the safety stock.

### CODE REVIEW

**Comment by LoisRForgeFlow** [comment]
@gurneyalex Sounds good, I will try to find some time to review this ASAP.

Thanks for the contribution!

**Review by SilvioC2C — APPROVED** [review]
LGTM

**Comment by gurneyalex** [comment]
@LoisRForgeFlow I updated the PR with some improvements to handle "low ressource" period for a MRP area (maybe missing employees, broken equipment...) when we could need to consume the safety stock without rebuilding it. 

Using this is optional.

**Review by LoisRForgeFlow — COMMENTED** [review]
Hi @gurneyalex 

Finally I'm reviewing this. The general idea of the prioritize safety stock is a good addition (as that it is very standard in other MRP implementations). 

With the secondary settings added, I'm more hesitant because they feel a bit ad-hoc, some thoughts:
- forces a specific date logic similar to a freezing/planning horizon but only for safety stock.
- Considers only weekly buckets planning.
- The date computed is the date to schedule the finish date, the trigger dates can be today if the LT is long enough. In other words, you are throwing procurement recommendations before that safety stock lead date.

The 2 first points are more comments than other thing, we can merge and let them evolve and improve if needed. For the 3rd one is more a warning and I want to know if that was intended, see below a couple of screenshots highlighting the problem:

image

image

**Comment by gurneyalex** [comment]
@LoisRForgeFlow thanks for the review :) 

Point 3: Good catch, I forgot to take the lead time into account. I'll fix this. 

Point 1 and 2: yes this is a bit of a hack and comes straight from my current customer's requirements. They do give some flexibility by being at the product.mrp.area level. I had originally proposed to have a single date on the mrp.multi.level wizard screen, but the customer has a case where for instance one specific machine is unavailable for 2 weeks and this makes a set of products "harder to produce" during the period, so we chose to move things to the mrp.area which gives a finer control. 

If you are more at ease with a single date (not computed) on the mrp.area I can change this PR to go in this direction and then possibly change the logic of computing this date in a customer specific module. What do you think?

File: mrp_multi_level/wizards/mrp_multi_level.py
Comment (SilvioC2C) [inline mrp_multi_level/wizards/mrp_multi_level.py]:
```suggestion
            product_mrp_area.mrp_area_id.priorize_safety_stock
            and date
            and product_mrp_area.mrp_area_id.safety_stock_target_date
            and date < product_mrp_area.mrp_area_id.safety_stock_target_date
```

Will fix: https://github.com/OCA/manufacture/actions/runs/7100988379/job/19328179577?pr=1123#step:8:587

File: mrp_multi_level/wizards/mrp_multi_level.py
Comment (SilvioC2C) [inline mrp_multi_level/wizards/mrp_multi_level.py]:
```suggestion
                and move.mrp_date
                and product_mrp_area.mrp_area_id.safety_stock_target_date
                and move.mrp_date > product_mrp_area.mrp_area_id.safety_stock_target_date
```

**Review by SilvioC2C — APPROVED** [review]
LGTM, tests are red because of failing dates comparison (some of them are not properly set so we get `date <operator> bool` which triggers a `TypeError`).

I'm suggesting changes but they may not be functionally ok (maybe we want to handle cases differently if one of the dates is not set?).

**Comment by LoisRForgeFlow** [comment]
> @LoisRForgeFlow thanks for the review :)
> 
> Point 3: Good catch, I forgot to take the lead time into account. I'll fix this.
> 
> Point 1 and 2: yes this is a bit of a hack and comes straight from my current customer's requirements. They do give some flexibility by being at the product.mrp.area level. I had originally proposed to have a single date on the mrp.multi.level wizard screen, but the customer has a case where for instance one specific machine is unavailable for 2 weeks and this makes a set of products "harder to produce" during the period, so we chose to move things to the mrp.area which gives a finer control.
> 
> If you are more at ease with a single date (not computed) on the mrp.area I can change this PR to go in this direction and then possibly change the logic of computing this date in a customer specific module. What do you think?

Since it is optinal to use it, I'm fine with it as it is now. If you just fix point 3, I think we could move forward

**Comment by LoisRForgeFlow** [comment]
@gurneyalex Also as a reminder, please do not merge PRs of mrp_multi_level, let us control the merges. I hope you understand, I might have approved a PR but maybe give a final thought before merging or maybe @JordiBForgeFlow is involved in a final review too.

**Comment by gurneyalex** [comment]
> please do not merge PRs of mrp_multi_leve

Understood. Sorry about this.

**Comment by gurneyalex** [comment]
> if you just fix point 3

It should be fixed. I'm adding some tests for this

**Comment by JordiBForgeFlow** [comment]
@gurneyalex I am confused about this feature. It's not the same as what is in the description of the PR right? You are introducing 2 features:

1. Rebuild now safety if stock level requires it -> I believe that this solves a bug. I propose that you remove the boolean and set it as the default behaviour.

2. Rebuild safety stock on a future date -> does not make sense to me, at least as part of the main module. Seems to me more like a customization. Besides, adding a specific date in the MRP area does not make sense. Because this would apply to all products in the area.

I propose that you introduce whatever hooks you need in this module that can help you to create an additional module that implements this feature.

**Comment by gurneyalex** [comment]
@JordiBForgeFlow Thanks for the feedback. I'll update the PR.

**Comment by gurneyalex** [comment]
The PR was updated. The code I split for the safety stock rebuilding is available on #1169

File: mrp_multi_level/wizards/mrp_multi_level.py
Comment (JordiBForgeFlow) [inline mrp_multi_level/wizards/mrp_multi_level.py]: This text is not relevant any more, right?

**Review by JordiBForgeFlow — APPROVED** [review]
:+1: Code review. Thanks for the fix!

File: mrp_multi_level/wizards/mrp_multi_level.py
Comment (gurneyalex) [reply mrp_multi_level/wizards/mrp_multi_level.py]: you're right. Fixed in d45721609ced96f652d83f02ec6c53d37e36b4c9

Happy new year to the ForgeFlow team :grin:

**Pull request merged**
